### ISSUE TITLE
Add Indexed Property for Class Creation

### ISSUE THREAD

**Issue by ianpward** [realm/realm-studio#948]
Add ability to create an indexed property when creating a class

**Comment by kraenhansen** [comment]
As far as I recall, what properties are indexed is a per device setting that is not saved to the Realm.

Perhaps Simon can confirm?

**Comment by cmelchior** [comment]
Yes, indexes are not synchronized but must be added manually on each device

**Comment by cmelchior** [comment]
It might still make sense to be able to add indexes on the serverside though.

**Comment by kneth** [comment]
I am a bit confused here. Isn't the information whether a column is indexed or not part of the schema? And isn't it synced? The actual index is not synced. But there is no sync instruction for setting an index on an existing colum, right? (mostly a question to @simonask)

**Comment by nirinchev** [comment]
The index information isn't synced - every Realm client can define their own indices.

**Comment by simonask** [comment]
Exactly. The rationale behind this is that especially with query-based sync, it is not unlikely that the user wants different indexes the server and on different clients (for instance, the server should have the indexes that best aid query-based sync, and these could be different from those required in mobile apps, but in turn the data adapter may not be interested in having any indexes at all).

Indexes are an optimization tool, but can very easily become a pessimization tool if misapplied.

It's clear that we need a way to control indexes on the server for query-based sync (today there is no way), but it should likely come as part of a more advanced server tuning toolset. For example, it would be nice to give the user an overview over which queries the server is spending time on, so the user can add indexes accordingly. This is already described in various issues that I haven't taken the time to hunt down (soz). :-)

**Comment by astigsen** [comment]
We probably want to find some way to at least specify default indexes, so that when people open read-only realms, they can still get an relevant indexes.

With js we are also seeing a lot of usage where people do not specify a schema, but get it straight from the file, so in that case it would also be nice for them to get default indexes.

**Comment by simonask** [comment]
From Sync's perspective, adding an index locally is a valid read-only operation. Sync's notion of "read-only" is different from Core. Since indexes are not represented in the Sync history format, nothing will be uploaded to the server if a client creates an index.

I don't believe the server should tell the client which indexes should exist. The server has no idea how that client plans to query the Realm.
